tron/kubernetes.py

```diff
@@ -148,35 +148,58 @@ def handle_event(self, event: Event) -> None:
-                    abnormal_exit = termination_metadata.get("exitCode") == 0 and (
-                        termination_metadata.get("finishedAt") is None and termination_metadata.get("reason") is None
-                    )
-                    if abnormal_exit:
-                        self.log.warning("Container never started due to a Kubernetes/infra flake!")
-                        self.log.warning(
-                            f"If automatic retries are not enabled, run `tronctl retry {self.id}` to retry."
-                        )
-
-            # -9 is EXIT_KUBERNETES_ABNORMAL from actionrun.py - we just can't import that value right now since there'd be a circular dependency
-            self.exited(-9 if abnormal_exit else 0)
-        elif k8s_type == "failed":
-            # TODO: we should reach into the container status field here and get the actual exit code
-            self.exited(1)
+                    state_termination_metadata = main_container_state.get("terminated", {}) or {}
+                    last_state_termination_metadata = main_container_last_state.get("terminated", {}) or {}
+                    if k8s_type == "finished":
+                        # this is kinda wild: we're seeing that a kubelet will sometimes fail to start a container (usually
+                        # due to what appear to be race conditons like those mentioned in
+                        # https://github.com/kubernetes/kubernetes/issues/100047#issuecomment-797624208) and then decide that
+                        # these Pods should be phase=Succeeded with an exit code of 0 - even though the container never actually
+                        # started. So far, we've noticed that when this happens, the finished_at and reason fields will be None
+                        # and thus we'll check for at least one of these conditions to detect an abnormal exit and actually "fail"
+                        # the affected action
+                        # NOTE: hopefully this won't change too drastically in future k8s upgrades without the actual problem (incorrect
+                        # success) being fixed :p
+                        if state_termination_metadata.get("exitCode") == 0 and (
+                            state_termination_metadata.get("finishedAt") is None
+                            and state_termination_metadata.get("reason") is None
+                        ):
+                            # -9 is EXIT_KUBERNETES_ABNORMAL from actionrun.py - we just can't import that value right now since there'd be a circular dependency
+                            exit_code = -9
+                            self.log.warning("Container never started due to a Kubernetes/infra flake!")
+                            self.log.warning(
+                                f"If automatic retries are not enabled, run `tronctl retry {self.id}` to retry."
+                            )
+                    elif k8s_type == "failed" or k8s_type == "killed":
```

Comment:
nit: 
```suggestion
                    elif k8s_type in {"failed", "killed"}:
```
(maybe also extract this to a named constant?)